pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go

```diff
@@ -159,6 +208,46 @@ func (m *Stub) Stop() error {
 	return m.cleanup()
 }
 
+func (m *Stub) Watch(kubeletEndpoint, resourceName, pluginSockDir string) {
+	for {
+		select {
+		// Detect a kubelet restart by watching for a newly created
+		// 'pluginapi.KubeletSocket' file. When this occurs, restart
+		// the device plugin server
+		case event := <-m.kubeletRestartWatcher.Events:
+			if event.Name == kubeletEndpoint && event.Op&fsnotify.Create == fsnotify.Create {
+				klog.InfoS("inotify: file created, restarting", "kubeletEndpoint", kubeletEndpoint)
+				var lastErr error
+
+				err := wait.PollUntilContextTimeout(context.Background(), 10*time.Second, 2*time.Minute, false, func(context.Context) (done bool, err error) {
+					restartErr := m.Restart()
+					if restartErr == nil {
+						return true, nil
+					}
+					klog.ErrorS(restartErr, "Retrying after error")
+					lastErr = restartErr
+					return false, nil
+				})
+				if err != nil {
```

Comment:
the two `return` statements above return `nil` as error.

when would this `err` be non-nil ?

Reply:
`PollUntilContextTimeout` polls until it returns true, an error, or hits a deadline.

The aim here is to allow time (at least 2 minute) for device plugin to restart. Please refer to this [ review comment](https://github.com/kubernetes/kubernetes/pull/118534#discussion_r1331931164) for the reason for adding the retry mechanism. We want to stop polling only if the plugin restart succeeds (and therefore exit with `true` and `nil` error). If the restart fails (or for some other reason) we return a non-nil error the polling would stop and we won't retry. Because we want to poll until the restart happens successfully and timeout only after 2 minutes if restart does not succeed we don't return a non-nil error.